src/firmware/shared/circular_buffer.c

```diff
@@ -0,0 +1,54 @@
+#include "firmware/shared/circular_buffer.h"
+
+void circular_buffer_init(struct circular_buffer *cbuff)
+{
+    cbuff->head   = 0;
+    cbuff->tail   = 0;
+    cbuff->isFull = false;
+    // Initialize buffer size to CIRCULAR_BUFFER_MAX_NUM_ELEMENTS
+    cbuff->bufferData = (float *)malloc(sizeof(float) * CIRCULAR_BUFFER_MAX_NUM_ELEMENTS);
+}
+
+void circular_buffer_free(struct circular_buffer *cbuff)
+{
+    free(cbuff->bufferData);
+}
+
+// Note: Implementation assumes head and tail are equal when full
+void circular_buffer_push(struct circular_buffer *cbuff, float data)
+{
+    cbuff->bufferData[cbuff->head % CIRCULAR_BUFFER_MAX_NUM_ELEMENTS] = data;
+    cbuff->head++;
+    if (cbuff->head > CIRCULAR_BUFFER_MAX_NUM_ELEMENTS)
```

Comment:
Depending on how quickly things are getting added to the buffer over time, when head is > int_max, it'll overflow and mess up the tail calculations. 

It will also mess up the bufferData index if the CIRCULAR_BUFFER_MAX_NUM_ELEMENTS is not a multiple of INT_MAX.

Most implementations of the circular buffers that I've seen use head and tail in the range [0, MAX_NUM_ELEMENTS] which solves this problem.

Reply:
Good to know about keeping it in that range. I'll have it be modified using modulus directly instead of doing it in the check.